File: src/main/java/com/example/chatService/service/FriendService.java
package com.example.chatService.service;

import com.example.chatService.component.WebSocketEventPublisher;
import com.example.chatService.dto.*;
import com.example.chatService.entity.Friend;
import com.example.chatService.entity.User;
import com.example.chatService.repository.FriendRepository;
import com.example.chatService.repository.FriendRequestRepository;
import com.example.chatService.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.*;

@Service
@RequiredArgsConstructor
@Transactional
@Slf4j
public class FriendService {

    private final UserRepository userRepository;
    private final FriendRepository friendRepository;
    private final WebSocketEventPublisher eventPublisher;

    @Transactional
    public FriendRequestResponseDto sendFriendRequest(String fromUserId, String toUserId) {
        User from = userRepository.findById(fromUserId)
                .orElseThrow(() -> new IllegalArgumentException("From user not found"));

        User to = userRepository.findById(toUserId)
                .orElseThrow(() -> new IllegalArgumentException("Target user not found"));

        if (from.getId().equals(to.getId())) {
            throw new IllegalArgumentException("자기 자신에게 보낼 수 없습니다");
        }

        if (friendRepository.existsByUserAndFriend(from, to) || friendRepository.existsByUserAndFriend(to, from)) {
            throw new IllegalArgumentException("이미 요청을 보냈거나 친구입니다.");
        }

        Friend f = Friend.builder()
                .user(from)
                .friend(to)
                .status(FriendStatus.PENDING)
                .build();

        friendRepository.save(f);

        PublishFriendEvent publishEvent = PublishFriendEvent.builder()
                                                .type(FriendEventType.FRIEND_REQUEST)
                                                .fromUserId(fromUserId)
                                                .fromUserNickname(from.getUsername()).build();

        eventPublisher.publishFriendEvent(toUserId, publishEvent);

        return FriendRequestResponseDto.builder()
                .fromUserId(fromUserId)
                .toUserId(toUserId)
                .status(f.getStatus().name())
                .build();
    }

    @Transactional
    public List<FriendRequestDto> getReceivedRequests(String userId) {
        List<Friend> requests = friendRepository.findPendingRequestsReceivedByUserId(userId);

        return requests.stream()
                .map(friends -> FriendRequestDto.builder()
                        .id(friends.getId())
                        .toUserId(friends.getFriend().getId())
                        .fromUserNickname(friends.getUser().getUsername())
                        .fromUserId(friends.getUser().getId())
                        .status(friends.getStatus())
                        .build())
                .toList();
    }

    @Transactional(readOnly = true)
    public List<User> getFriendList(String userId) {
        return friendRepository.findAllRelatedWithFetch(userId)
                .stream()
                .filter(f -> f.getStatus() == FriendStatus.ACCEPTED)
                .map(f -> f.getUser().getId().equals(userId) ? f.getFriend() : f.getUser())
                .distinct()
                .toList();
    }

    @Transactional
    public void removeFriend(String me, String target) {

        User userA = userRepository.findById(me)
                .orElseThrow(() -> new IllegalArgumentException("user not found, me = "  + me + " target id = " + target));
        User userB = userRepository.findById(target)
                .orElseThrow(() -> new IllegalArgumentException("user not found, me = "  + me + " target id = " + target));

        List<Friend> friends = friendRepository.findFriendRelation(userA, userB);

        friendRepository.deleteAll(friends);
    }
    @Transactional
    public String acceptFriendRequest(Long requestId) {
        Friend request = friendRepository.findById(requestId)
                .orElseThrow(() -> new IllegalArgumentException("요청을 찾을 수 없습니다"));

        if ( request.getStatus() != FriendStatus.PENDING) {
            throw new IllegalStateException("이미 처리된 요청입니다");
        }

        User sender = request.getFriend(); // 요청 보낸사람
        User receiver = request.getUser(); // 요청 받은사람

        request.setStatus(FriendStatus.ACCEPTED);
        friendRepository.save(request);

        Friend friendA = Friend.builder()
                .user(sender)
                .friend(receiver)
                .status(FriendStatus.ACCEPTED)
                .build();

        Friend friendB = Friend.builder()
                .user(receiver)
                .friend(sender)
                .status(FriendStatus.ACCEPTED)
                .build();

        friendRepository.save(friendA);
        friendRepository.save(friendB);

        PublishAcceptFriendEvent publishAcceptFriendEvent =
                        PublishAcceptFriendEvent.builder()
                            .type(FriendEventType.FRIEND_ACCEPT)
                            .friendId(receiver.getId()).build();

        eventPublisher.publishAcceptFriendEvent(
                            sender.getId(),
                            publishAcceptFriendEvent);

        return "친구 신청 수락 완료";
    }

    @Transactional
    public String reject(Long requestId) {
        Friend req = friendRepository.findById(requestId)
                .orElseThrow(() -> new IllegalArgumentException("요청 없음"));

        if (req.getStatus() != FriendStatus.PENDING)
            throw new IllegalStateException("이미 처리됨");

        req.setStatus(FriendStatus.BLOCKED);
        return "거절 완료";
    }
}

File: src/main/java/com/example/chatService/service/FriendService.kt
package com.example.chatService.service

import com.example.chatService.component.WebSocketEventPublisher
import com.example.chatService.dto.*
import com.example.chatService.entity.Friend
import com.example.chatService.entity.User
import com.example.chatService.repository.FriendRepository
import com.example.chatService.repository.UserRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class FriendService (
        private val userRepository: UserRepository,
        private val friendRepository: FriendRepository,
        private val eventPublisher: WebSocketEventPublisher
){

    @Transactional
    fun sendFriendRequest(fromUserId : String, toUserId : String) : FriendRequestResponseDto {
        val from = userRepository.findById(fromUserId)
                .orElseThrow { IllegalArgumentException("from user not found") }

        val to = userRepository.findById(toUserId)
                .orElseThrow { IllegalArgumentException("Target user not found")}

        if (from.id.equals(to.id)) {
            throw IllegalArgumentException("자기 자신에게 보낼 수 없습니다.")
        }

        if (friendRepository.existsByUserAndFriend(from, to) || friendRepository.existsByUserAndFriend(to, from)) {
            throw IllegalArgumentException("이미 요청을 보냈거나 친구입니다.")
        }

        val f = Friend(
                user = from,
                friend = to,
                status = FriendStatus.PENDING
        )
        friendRepository.save(f)

        val publishEvent = PublishFriendEvent(
                type = FriendEventType.FRIEND_REQUEST,
                fromUserId = fromUserId,
                fromUserNickname = from.usernameValue
        )

        eventPublisher.publishFriendEvent(toUserId, publishEvent)

        return FriendRequestResponseDto(
                fromUserId = fromUserId,
                toUserId = toUserId,
                status = f.status!!.name
        )
    }

    @Transactional
    fun getReceivedRequests(userId : String) : List<FriendRequestDto> {
        val request = friendRepository.findPendingRequestsReceivedByUserId(userId)

        return request.map {
            friends -> FriendRequestDto(
                    id = friends.id,
                    toUserId = friends.friend?.id,
                    fromUserNickname = friends.user!!.usernameValue,
                    fromUserId = friends.user?.id,
                    status = friends.status
            )
        }.toList()
    }

    @Transactional(readOnly = true)
    fun getFriendList(userId: String) : List<User?> {
        return friendRepository.findAllRelatedWithFetch(userId)
                .filter { f -> f.status == FriendStatus.ACCEPTED }
                .map { if (it.user?.id == userId) it.friend else it.user  }
                .distinct()
    }

    @Transactional
    fun removeFriend(me : String, target: String) {
        val userA = userRepository.findById(me)
                .orElseThrow { IllegalArgumentException("user not found, me = $me target id = $target") }
        val userB = userRepository.findById(target)
                .orElseThrow { IllegalArgumentException("user not found, me = $me target id = $target") }

        val friends = friendRepository.findFriendRelation(userA, userB)

        friendRepository.deleteAll(friends)
    }

    @Transactional
    fun acceptFriendRequest(requestId : Long) : String {
        val request = friendRepository.findById(requestId)
                .orElseThrow { IllegalArgumentException("요청을 찾을 수 없습니다.") }

        if ( request.status != FriendStatus.PENDING) {
            throw IllegalArgumentException("이미 처리된 요청입니다.")
        }

        val sender = request.friend
                ?: throw IllegalArgumentException("요청 보낸 사용자가 없습니다")
        val receiver = request.user
                ?: throw IllegalArgumentException("요청 받은 사용자가 없습니다")

        request.status = FriendStatus.ACCEPTED
        friendRepository.save(request)

        val friendA =  Friend(
                user = sender,
                friend = receiver,
                status = FriendStatus.ACCEPTED
        )

        val friendB = Friend(
                user = receiver,
                friend = sender,
                status = FriendStatus.ACCEPTED
        )

        friendRepository.save(friendA)
        friendRepository.save(friendB)

        val publishAcceptFriendEvent = PublishAcceptFriendEvent(
                type = FriendEventType.FRIEND_ACCEPT,
                friendId = receiver.id
        )

        eventPublisher.publishAcceptFriendEvent(
                sender.id,
                publishAcceptFriendEvent
        )
        return "친구 수락 완료"
    }

    @Transactional
    fun reject(requestId: Long) : String {
        val req = friendRepository.findById(requestId)
                .orElseThrow { IllegalArgumentException("요청 없음") }

        if (req.status != FriendStatus.PENDING)
            throw IllegalArgumentException("이미 처리됨")

        req.status = FriendStatus.BLOCKED
        return "거절 완료"
    }
}